### PR TITLE
Override declared adapter at runtime

### PR DESCRIPTION
**Pull request by totomz**
With this change it is possible to
- override, at runtime, the declared adapter for a class.
- register an adapter for a final class that declared in a library, for which is impossible to add an annotation at compile time.

The usecase is the same defined in the test: a class (`SimpleObj`) that is persisted using a dto (`SimpleDto`); so `SimpleObj` is annotated with `@Banana(adapter = MagicAdapter.class)`. 
We can consider `SimpleDto` as a class declared in an external library

In this case, `BananaRama.using()` overrides only the first adapter, so it is not a working alternative

### CODE REVIEW

File: src/main/java/org/bananarama/BananaRama.java [hunk 1]
```diff
@@ -57,8 +63,18 @@ private void init(){
...
     private  Adapter getAdapterForClass(Class<?> clazz){
-        if(!clazz.isAnnotationPresent(Banana.class))    
+        
+        if(customAdapters.containsKey(clazz)) {
+            return getAdapter(customAdapters.get(clazz));
+        }
+        
+        if(!clazz.isAnnotationPresent(Banana.class))    {
+            if(clazz.getName().equals("com.google.cloud.datastore.Entity")) {
+                System.out.println("ASDASDSADSADA");
```
Comment (cr0wbar) [inline src/main/java/org/bananarama/BananaRama.java]: This is unacceptable

File: src/main/java/org/bananarama/BananaRama.java [hunk 2]
```diff
@@ -41,6 +42,11 @@
     
     private final StripedLock slock;
     private final Map<Class<?>, Adapter<?>> adapters;
+    private static final Map<Class<?>, Class<? extends Adapter>> customAdapters = new ConcurrentHashMap<>();
```
Comment (cr0wbar) [inline src/main/java/org/bananarama/BananaRama.java]: static modifier does not belong here.

**Comment by totomz** [comment]
uuuups, did not clean up 😄

File: src/main/java/org/bananarama/BananaRama.java [hunk 3]
```diff
@@ -41,6 +42,11 @@
...
+    
+    public static void registerAdapter(Class<? extends Adapter> adapter, Class forEntiy) {
```
Comment (cr0wbar) [inline src/main/java/org/bananarama/BananaRama.java]: BAD IDEA. This can easily generate unpredictable behavior at runtime as it breaks correlation between how class are annotated and how they are loaded.

Comment (cr0wbar) [inline src/main/java/org/bananarama/BananaRama.java]: static modifer is bullshit

File: src/main/java/org/bananarama/BananaRama.java [hunk 4]
```diff
@@ -42,10 +42,18 @@
...
     
-    public static void registerAdapter(Class<? extends Adapter> adapter, Class forEntiy) {
+    /**
+     * Map the specified {@link Adapter} for a given entity. 
+     * This method is usefull to allow final class from external libraries to be managed by BananaRama
+     * @param adapter
+     * @param forEntiy 
+     * @return this instance
+     */
+    public BananaRama registerAdapter(Class<? extends Adapter> adapter, Class forEntiy) {
```
Comment (cr0wbar) [inline src/main/java/org/bananarama/BananaRama.java:54]: Ok better, but again, same problem, this implies that annotated class are not necessarily loaded as expected which is very dangerous.

File: src/test/java/magic/cascade/ForceCustomAdaptersTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+               Optional.ofNullable(SimpleObj.class.getDeclaredAnnotation(MapWith.class))
+                    .map(MapWith::value)
+                    .get());
+
+        // SimpleMapper map to SimpleDto
+        Assert.assertNotNull(SimpleMapper.class.getDeclaredMethod("toObj", SimpleDto.class));
+       
+        // SimpleDto is bananaed using a ListAdapter
+        Assert.assertEquals(
+               ListAdapter.class,
```
Comment (cr0wbar) [inline src/test/java/magic/cascade/ForceCustomAdaptersTest.java]: Split code to make it more readable. I can't see shit.

File: src/main/java/org/bananarama/BananaRama.java (hunk 4 above)
Comment (totomz) [inline src/main/java/org/bananarama/BananaRama.java:54]: Yep, but this is the required behaviour....

File: src/test/java/magic/cascade/MemoryAdapter.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import org.bananarama.crud.ReadOperation;
+import org.bananarama.crud.UpdateOperation;
+
+/**
+ * Simple adapter that store data in a public static map
+ * @author Tommaso Doninelli <tommaso.doninelli@gmail.com>
+ */
+public class MemoryAdapter implements Adapter<Object>{
+    
+    protected static final List<Object> datastore = Collections.synchronizedList(new ArrayList<>());
```
Comment (cr0wbar) [inline src/test/java/magic/cascade/MemoryAdapter.java:21]: What's with this static stuff??? Remove it.

File: src/test/java/magic/cascade/MemoryAdapter.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        return new CreateOperation<T>() {
+            @Override
+            public CreateOperation<T> from(Stream<T> data) {
+                data.forEach(datastore::add);
+                return this;
+            }
+
+            @Override
+            public CreateOperation<T> from(Stream<T> data, QueryOptions options) {
+                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (cr0wbar) [inline src/test/java/magic/cascade/MemoryAdapter.java]: Generated comment. BAD!

File: src/test/java/magic/cascade/MemoryAdapter.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    public <T> ReadOperation<T> read(Class<T> clazz) {
+        return new ReadOperation<T>() {
+            @Override
+            public Stream<T> all() {
+                return (Stream<T>) datastore.stream();
+            }
+
+            @Override
+            public Stream<T> all(QueryOptions options) {
+                throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (cr0wbar) [inline src/test/java/magic/cascade/MemoryAdapter.java]: CLEAN THE CODE!!

File: src/test/java/magic/cascade/MemoryAdapter.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            }
+
+            @Override
+            public void close() throws IOException {}
+        };
+    }
+
+    @Override
+    public <T> UpdateOperation<T> update(Class<T> clazz) {
+        throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (cr0wbar) [inline src/test/java/magic/cascade/MemoryAdapter.java]: OMG, Seriously??!!

File: src/test/java/magic/cascade/ForceCustomAdaptersTest.java (hunk 5 above)
Comment (totomz) [inline src/test/java/magic/cascade/ForceCustomAdaptersTest.java]: miii che cagacazzi.....ma è leggibilissimo!

File: src/test/java/basic/ListAdapter.java [hunk 10]
```diff
@@ -35,11 +35,17 @@
 
     private final List<Object> backend = new ArrayList<>();
 
+    public ListAdapter() {
+        System.out.println("uallaualla");
```
Comment (cr0wbar) [inline src/test/java/basic/ListAdapter.java]: Poor coding

File: src/test/java/basic/ListAdapter.java [hunk 11]
```diff
@@ -35,11 +35,17 @@
...
+    }
+    
+    
+    
     @Override
     public <T> CreateOperation<T> create(Class<T> clazz) {
        return new CreateOperation<T>() {
            @Override
-           public CreateOperation<T> from(Stream<T> data) {
+           public CreateOperation<T> from(Stream<T> data) {
```
Comment (cr0wbar) [inline src/test/java/basic/ListAdapter.java:42]: Unnecessary change

File: src/main/java/org/bananarama/BananaRama.java [hunk 12]
```diff
@@ -41,6 +42,19 @@
     
     private final StripedLock slock;
     private final Map<Class<?>, Adapter<?>> adapters;
+    private final Map<Class<?>, Class<? extends Adapter>> customAdapters = new ConcurrentHashMap<>();
```
Comment (cr0wbar) [inline src/main/java/org/bananarama/BananaRama.java:45]: Inline assignment not consistent with rest of code.

File: src/test/java/magic/cascade/MemoryAdapter.java (hunk 6 above)
Comment (totomz) [inline src/test/java/magic/cascade/MemoryAdapter.java:21]: It is required for the test. It is not an adapter, is a simple test case

File: src/test/java/magic/cascade/MemoryAdapter.java (hunk 7 above)
Comment (totomz) [inline src/test/java/magic/cascade/MemoryAdapter.java]: Yep, it is not a working code. It is a fake (in the meaning of FAKE) adapter, needed only for the test

File: src/test/java/basic/ListAdapter.java (hunk 10 above)
Comment (totomz) [inline src/test/java/basic/ListAdapter.java]: uh, sorry again for this

**Comment by cr0wbar** [comment]
I do not agree with this new feature. if you want to use a specific adapter at runtime just run
`Adapter adap  = Bananarama.using(MyAdapter.class)` and then run all operation explicitly using the adapter. This should work even if the class is final. This way you will be guaranteed to use always the same instance. No need to modify the inner workings of `BananaRama` just for a specific case, which seems to be easily supported anyway.

File: src/test/java/magic/cascade/MemoryAdapter.java [hunk 13]
```diff
@@ -31,7 +31,7 @@
 
             @Override
             public CreateOperation<T> from(Stream<T> data, QueryOptions options) {
```
Comment (cr0wbar) [inline src/test/java/magic/cascade/MemoryAdapter.java:33]: This adapter is a stub at best. It is not acceptable.

Comment (totomz) [inline src/test/java/magic/cascade/MemoryAdapter.java:33]: It **is** a stub, needed only for a specific test (this is why it is in `src/test` )

**Comment by totomz** [comment]
Ok for closing the pr now, I'll add here an example of the usecase to show my point :smile:
